awx/ui_next/src/util/qs.test.js

```diff
@@ -810,4 +818,66 @@ describe('qs (qs.js)', () => {
       });
     });
   });
+
+  describe('replaceNamespacedParams', () => {
+    const config = {
+      namespace: 'template',
+      defaultParams: { page: 1, page_size: 5, order_by: 'name' },
+      integerFields: ['page'],
+    };
+
+    test('should update namespaced param', () => {
+      const query = 'template.name__icontains=workflow&template.page=2';
+      const newParams = {
+        page: 3,
+      };
+      expect(replaceNamespacedParams(config, query, newParams)).toEqual(
+        'template.name__icontains=workflow&template.page=3'
+      );
+    });
+
+    test('should add new namespaced param', () => {
+      const query = 'template.name__icontains=workflow&template.page=2';
+      const newParams = {
+        or__type: 'job_template',
+      };
+      expect(replaceNamespacedParams(config, query, newParams)).toEqual(
+        'template.name__icontains=workflow&template.or__type=job_template&template.page=2'
+      );
+    });
+
+    test('should maintain non-namespaced param', () => {
+      const query = 'foo=bar&template.page=2&template.name__icontains=workflow';
+      const newParams = {
+        page: 3,
+      };
+      expect(replaceNamespacedParams(config, query, newParams)).toEqual(
+        'foo=bar&template.name__icontains=workflow&template.page=3'
+      );
+    });
+
+    test('should omit null values', () => {
+      const query = 'template.name__icontains=workflow&template.page=2';
+      const newParams = {
+        page: 3,
+        name__icontains: null,
+      };
+      expect(replaceNamespacedParams(config, query, newParams)).toEqual(
+        'template.page=3'
+      );
+    });
+
+    // This fix needed after we're confident refactoring components
+    // to use replaceNamespacedParams provides equivalent functionality
+    test.skip('should not alter params of other namespaces', () => {
+      const query =
+        'template.name__icontains=workflow&template.page=2&credential.page=3';
+      const newParams = {
+        page: 3,
+      };
+      expect(replaceNamespacedParams(config, query, newParams)).toEqual(
+        'template.name__icontains=workflow&template.page=3&credential.page=3'
```

Comment:
On first glance, the updates to replace params including the null-valued keys which should be removed is a bug...but it seems you are testing that with this new replaceNamespacedParams and it seems to be working fine.  That again leads me to believe you should just make the old function non-exported and remove those tests as what they were intending to cover is done with these new ones.

I also feel that you should have a test which makes sure default params are stripped out and not encoded (that's more of a concern at the pushHistoryState level, which was

Reply:
Yeah, I agree about not exporting the old function. It's still used in `ActivityStream` though, so I need to see if a different more focused util would do the job there. I also use the old utils inside `replaceNamespacedParams`, so I will keep it and its tests around, but will see about removing the export